go/saltpack/decrypt.go

```diff
@@ -92,14 +92,30 @@ func (ds *decryptStream) read(b []byte) (n int, err error) {
 	return n, nil
 }
 
-func (ds *decryptStream) readHeader() error {
-	var hdr EncryptionHeader
-	seqno, err := ds.mps.Read(&hdr)
+func (ds *decryptStream) readHeader(rawReader io.Reader) error {
+	// Parse the length of the header.
+	var headerLength int
+	seqno, err := ds.mps.Read(&headerLength)
+	if err != nil {
+		return ErrFailedToDecodeHeaderLength
+	}
+	// Read the header bytes.
+	headerBytes := make([]byte, headerLength)
+	_, err = io.ReadFull(rawReader, headerBytes)
```

Comment:
You know what would be cleaner?  You might hate this, but to msgpack the header twice!  That is, output `msgpack(msgpack(hdr))`.  Which is basically what you're doing.  But that would solve this problem, that you're relying on msgpack not buffering.  (BTW if you know it wouldn't buffer and read past the end, you could rely on my gimmick below to ensure it consumed all of the buffer).